Automattic/wp-calypso
# composite-checkout: Allow `isCompleteCallback` to be async

There are three related changes in this PR. 

#### Allow isCompleteCallback to be async

First, it modifies the `isCompleteCallback` property of Step Objects and Payment Method Objects to optionally return a Promise. If a Promise is returned, the step will be marked complete or incomplete when the Promise resolves, rather than immediately. This allows for asynchronous validation which will be helpful when using the server-side validation introduced in #38729 but it does create one complication; we don't want to trigger async validation on every render, which is how the completion function is triggered now. That could be very expensive if the validation is a server call.

#### Always enable Continue buttons

This leads to the second change, which has been a UX goal in itself. Currently the "Continue" button on each step is disabled while the step is incomplete. This isn't ideal since the disabled state doesn't give any information about what's wrong if `isCompleteCallback` returns false. In this PR we modify the button so that it's always enabled, but only goes to the next step if the callback returns true. To do this we move calling the callback until the button is pressed which resolves the issue of async validation mentioned above. 

This change also help simplify some areas of the code because we now store the complete status of a step as component state rather than part of the annotated step objects. Previously in order to make sure the `isCompleteCallback` had the most recently updated data we had to force re-renders on every data change to force it to be called. Now the re-render happens automatically when the state changes.

Note that if the step is incomplete and the Continue button is pressed, it will appear that nothing happens; it is up to the `isCompleteCallback` to display appropriate validation errors. This gives the callback complete control over how best to display validation errors (notifications, displaying things inline, etc.) but is not ideal because it's too easy to _not_ notify the user; suggestions are welcome about better ways to handle the situation.

#### Removing paymentData

This PR removes passing both `paymentData` and `activeStep` as arguments to `isCompleteCallback` and `isEditableCallback`. Those arguments were mostly unused and complicated both the implementation and the documentation of those functions.

Since we no longer have `paymentData` we also remove the `usePaymentData` hook which was introduced as a simple way to store arbitrary data in payment methods and step objects. However, in practice it is far too inflexible to use for real data. Removing it should help to make it clear that the data layer of your custom components doesn't matter as far as this API is concerned (although `useRegisterStore` remains and option).

#### Testing instructions

- Run calypso with `ENABLE_FEATURES=composite-checkout-wpcom npm start`.
- Add a product to your cart and visit checkout.
- Make sure all steps display correctly.
- Make sure "Continue" button on each step is always enabled.
- Make sure pressing "Continue" button if step is incomplete does not move to the next step.
- Make sure pressing "Continue" button if step is incomplete displays some form of validation error.
- Make sure submit payment button is disabled until the final step.
- Make sure submit payment button is enabled on final step.
